tools/database/src/memtrie.rs

```diff
@@ -352,3 +356,188 @@ impl FindBoundaryAccountCommand {
+
+        // Get the shard layout and ensure the protocol version upgrade actually
+        // contains a resharding.
+        let shard_layout =
+            epoch_manager.get_shard_layout_from_protocol_version(self.protocol_version);
+        let prev_shard_layout =
+            epoch_manager.get_shard_layout_from_protocol_version(self.protocol_version - 1);
+        assert_ne!(
+            shard_layout, prev_shard_layout,
+            "The provided protocol version does not contain a resharding."
+        );
+
+        println!("Searching for the first epoch of the provided protocol version");
+        // Find the fist epoch of the provided protocol version.
+        let head = store.chain_store().head()?;
+        let mut epoch_id = head.epoch_id;
+        loop {
+            let epoch_start_height = epoch_manager.get_epoch_start_from_epoch_id(&epoch_id)?;
+            let epoch_start_block_header =
+                store.chain_store().get_block_header_by_height(epoch_start_height)?;
+
+            let prev_block_hash = epoch_start_block_header.prev_hash();
+            let prev_epoch_id = epoch_manager.get_epoch_id(prev_block_hash)?;
+            let prev_protocol_version = epoch_manager.get_epoch_protocol_version(&prev_epoch_id)?;
+
+            if prev_protocol_version < self.protocol_version {
+                break;
+            }
+
+            epoch_id = prev_epoch_id;
+        }
+        println!("Found epoch id {epoch_id:?}");
+
+        // Find the resharding block. It is the last block of the previous epoch.
+        let epoch_start_height = epoch_manager.get_epoch_start_from_epoch_id(&epoch_id)?;
+        let epoch_start_block_hash =
+            store.chain_store().get_block_hash_by_height(epoch_start_height)?;
+        let epoch_start_block = store.chain_store().get_block(&epoch_start_block_hash)?;
+        let resharding_block_hash = epoch_start_block.header().prev_hash();
+        let resharding_block = store.chain_store().get_block(resharding_block_hash)?;
+        println!("Found resharding block {resharding_block_hash:?}");
+
+        // Prepare the resharding split shard params.
+        let resharding_block_info = BlockInfo {
+            hash: *resharding_block.hash(),
+            height: resharding_block.header().height(),
+            prev_hash: *resharding_block.header().prev_hash(),
+        };
+        let resharding_event =
+            ReshardingEventType::from_shard_layout(&shard_layout, resharding_block_info)?;
+        let resharding_event = resharding_event.expect("Could not create the resharding event");
+        let ReshardingEventType::SplitShard(split_shard_params) = resharding_event;
+        println!("Prepared split shard params: {split_shard_params:?}");
+
+        let shard_tries = runtime.get_tries();
+        if self.check_only {
+            Self::check_trie(
+                &shard_layout,
+                &shard_tries,
+                &resharding_block,
```

Comment:
Did we check if there is data loss for resharding_block +- 1 ? Overall, looks good.